### PR TITLE
Add events workflow

### PR DESCRIPTION
**Pull request by La0**
Migration from release-services

:warning: We should ship a first [libmozevent](https://github.com/La0/libmozevent/) version from a Mozilla repo to Pypi.

### CODE REVIEW

**Comment by La0** [comment]
This is now running OK on testing, replacing the release-services instance as https://eventlistener.testing.moz.tools/

This [demo build](https://phabricator-dev.allizom.org/harbormaster/build/2989/) on phab-dev was processed by this new instance

File: events/docker/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,9 @@
+FROM python:3
+
+ADD events /src
+
+RUN cd /src && pip install -r requirements.txt && python setup.py install
```
Comment (marco-c) [inline events/docker/Dockerfile]: Use `--disable-pip-version-check --no-cache-dir` for pip

File: events/docker/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,9 @@
+FROM python:3
```
Comment (marco-c) [inline events/docker/Dockerfile]: Can we use a slim image?

File: events/docker/bootstrap.sh [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+apt-get update
+apt-get install --no-install-recommends -y curl python2-minimal python-bz2file python-dev
+
+# Setup mercurial from its own website, without install pip2 which has a lot of dependencies
+curl -L https://www.mercurial-scm.org/release/mercurial-$MERCURIAL_VERSION.tar.gz | tar -C /opt -xvz
+cd /opt/mercurial-$MERCURIAL_VERSION
+python2 setup.py install
+
+# Setup mercurial with needed extensions
+hg clone https://hg.mozilla.org/hgcustom/version-control-tools /src/version-control-tools/
```
Comment (marco-c) [inline events/docker/bootstrap.sh]: Clone a specific revision of vct, so the build is replicable

File: events/docker/bootstrap.sh [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+mkdir -p /etc/mercurial/hgrc.d
+ln -s /src/docker/hgrc /etc/mercurial/hgrc.d/code-review.rc
+
+# Cleanup
```
Comment (marco-c) [inline events/docker/bootstrap.sh]: Apply the same cleanup steps as https://github.com/mozilla/code-coverage/blob/8c5de7b235cea360c8ad190e51459c5758768ede/bot/ci/bootstrap.sh to have a smaller image.

File: events/setup.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
+# -*- coding: utf-8 -*-
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import setuptools
+
+
+def read_requirements(file_):
```
Comment (marco-c) [inline events/setup.py]: We probably don't need the additional complexity in this, as we have no lines starting with "-e" or "http" or "https". You could probably do like https://github.com/mozilla/bugbug/blob/8b59321c5f694788606d7a07008238057292a38f/setup.py#L13.

File: events/code_review_events/cli.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    return parser.parse_args()
+
+
+class Events(object):
+    """
+    Listen to HTTP notifications from phabricator and trigger new try jobs
+    """
+
+    def __init__(self, cache_root):
+        # Create message bus shared amongst process
```
Comment (marco-c) [inline events/code_review_events/cli.py]:
```suggestion
        # Create message bus shared amongst processes
```

File: events/code_review_events/cli.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+            api_key=taskcluster_config.secrets["PHABRICATOR"]["api_key"],
+            url=taskcluster_config.secrets["PHABRICATOR"]["url"],
+            publish=taskcluster_config.secrets["PHABRICATOR"].get("publish", False),
+            risk_analysis_reviewers=taskcluster_config.secrets.get(
+                "risk_analysis_reviewers", []
+            ),
+        )
+        self.workflow.register(self.bus)
+
+        # Build mercurial worker & queue
```
Comment (marco-c) [inline events/code_review_events/cli.py]: Nit: ```suggestion
        # Build mercurial worker and queue
```

File: events/code_review_events/cli.py [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+        )
+        self.mercurial.register(self.bus)
+
+        # Create web server
+        self.webserver = WebServer(QUEUE_WEB_BUILDS)
+        self.webserver.register(self.bus)
+
+        # Setup monitoring for newly created tasks
+        self.monitoring = Monitoring(
+            QUEUE_MONITORING, taskcluster_config.secrets["admins"], 7 * 3600
```
Comment (marco-c) [inline events/code_review_events/cli.py]: Nit: put 7 * 3600 in a constant at the top of the file

File: events/code_review_events/cli.py [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+                self.bus.run(self.workflow.publish_results, QUEUE_PHABRICATOR_RESULTS)
+            )
+
+        # Start the web server in its own process
+        self.webserver.start()
+
+        # Run all tasks concurrently
+        run_tasks(consumers)
+
+        # Stop the webserver when other async process are stopped
```
Comment (marco-c) [inline events/code_review_events/cli.py]:
```suggestion
        # Stop the webserver when other async processes are stopped
```

File: events/code_review_events/workflow.py [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+                    logger.warning(
+                        "Failed to load build details", build=str(build), error=str(e)
+                    )
+                    continue
+
+                # Then send the build toward next stage
+                logger.info("Send build to Mercurial", build=str(build))
+                await self.bus.send(QUEUE_MERCURIAL, build)
+
+                # Report public bug as working
```
Comment (marco-c) [inline events/code_review_events/workflow.py]:
```suggestion
                # Report public bug as 'working' (in progress)
```

**Comment by La0** [comment]
Fixed all the comments + moved `Events` in `workflow.py`.
I just pushed on testing, will merge after some tests online

**Pull request merged**